app/models/manageiq/providers/vmware/discovery.rb

```diff
@@ -4,70 +4,59 @@ module ManageIQ
   module Providers
     module Vmware
       class Discovery
-        SERVER_CONSOLE_PORTS = [902, 912].freeze
-
-        ESX_PORTS = [902, 903].freeze
-
-        VC_PORTS = [
-          [
-            135, # VC < 5.1 or
-            7444 # VC >= 5.1
-          ],
-          [
-            139,  # VC < 5.1 or
-            2012, # VC >= 5.1
-            2013,
-            2014
-          ]
-        ].freeze
-
         def self.probe(ost)
-          # Check if VMware Server
-          ost.hypervisor << :vmwareserver if ManageIQ::NetworkDiscovery::Port.scan_open(ost, SERVER_CONSOLE_PORTS).length == 2
+          # Get Info from VMware (vSphere) Web Service API
+          info = retrieve_webservice_info(ost.ipaddr)
 
-          # First check if we can access the VMware webservice before even trying the port scans.
-          begin
-            require 'VMwareWebService/MiqVimClientBase'
-            MiqVimClientBase.new(ost.ipaddr, "test", "test")
-          rescue => err
-            $log&.debug("Vmware::Discovery: Failed to connect to VMware webservice: #{err}. ip = #{ost.ipaddr}")
-            return
-          end
+          hvisor_type = hypervisor_type(ost.ipaddr, info&.productLineId)
 
-          $log&.debug("Vmware::Discovery: ip = #{ost.ipaddr}, Connected to VMware webservice. Machine is either ESX or VirtualCenter.")
+          if ost.discover_types&.include?(hvisor_type)
+            ost.hypervisor << hvisor_type
+            ost.os << hypervisor_os_type(info&.osType)
```

Comment:
Probably would be good to add `unless info.osType.nil?` here so we don't potentially have `ost.os = [nil]` and cause the caller issues

Reply:
ost.os should result to be :mswin or :linux only 

I think it is better option than result ost.hypervisor.size != ost.os.size

Reply:
That's fair, osType is a required field in AboutInfo so it _shouldn't_ be nil

Reply:
I still don't see where is the problem - how should be possible to have `ost.os = [nil]` with this code?

If osType will be nil, then hypervisor_os_type returns :linux

Reply:
Ah was getting ahead of myself, I was thinking `hypervisor_os_type` should check if osType is nil instead of just defaulting to `:linux` but then `ost.os` would be `[nil]` so thought better to check if nil before calling it so `ost.os = [nil]` was an intermediate step in my head that I didn't write down haha

Reply:
Ok, now I'm confused what is your recommendation :)) 
If I check if osType isn't nil instead of defaulting to :linux, then `ost.hypervisor` and `ost.os` arrays can have different sizes. It doesn't cause error, but I don't know if next code is ok with that

Reply:
It looks like it just won't save the operating_system automatically https://github.com/ManageIQ/manageiq/blob/master/app/models/host.rb#L994
If you want to just have it default to linux that's fine